crates/matrix-sdk-common/src/linked_chunk/builder.rs

```diff
@@ -0,0 +1,486 @@
+/// error).
+#[allow(missing_debug_implementations)]
+pub struct LinkedChunkBuilder<const CAP: usize, Item, Gap> {
+    /// Work-in-progress chunks.
+    chunks: BTreeMap<ChunkIdentifier, TemporaryChunk<Item, Gap>>,
+
+    /// Is the final `LinkedChunk` expected to include an update history, as if
+    /// it were created with [`LinkedChunk::new_with_update_history`]?
+    build_with_update_history: bool,
+}
+
+impl<const CAP: usize, Item, Gap> Default for LinkedChunkBuilder<CAP, Item, Gap> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<const CAP: usize, Item, Gap> LinkedChunkBuilder<CAP, Item, Gap> {
+    /// Create an empty [`LinkedChunkBuilder`] with no update history.
+    pub fn new() -> Self {
+        Self { chunks: Default::default(), build_with_update_history: false }
+    }
+
+    /// Stash a gap chunk with its content.
+    ///
+    /// This can be called even if the previous and next chunks have not been
+    /// added yet. Resolving these chunks will happen at the time of calling
+    /// [`LinkedChunkBuilder::build()`].
+    pub fn push_gap(
+        &mut self,
+        previous: Option<ChunkIdentifier>,
+        id: ChunkIdentifier,
+        next: Option<ChunkIdentifier>,
+        content: Gap,
+    ) {
+        let chunk = TemporaryChunk { id, previous, next, content: ChunkContent::Gap(content) };
+        self.chunks.insert(id, chunk);
+    }
+
+    /// Stash an item chunk with its contents.
+    ///
+    /// This can be called even if the previous and next chunks have not been
+    /// added yet. Resolving these chunks will happen at the time of calling
+    /// [`LinkedChunkBuilder::build()`].
+    pub fn push_items(
+        &mut self,
+        previous: Option<ChunkIdentifier>,
+        id: ChunkIdentifier,
+        next: Option<ChunkIdentifier>,
+        items: impl IntoIterator<Item = Item>,
+    ) {
+        let chunk = TemporaryChunk {
+            id,
+            previous,
+            next,
+            content: ChunkContent::Items(items.into_iter().collect()),
+        };
+        self.chunks.insert(id, chunk);
+    }
+
+    /// Request that the resulting linked chunk will have an update history, as
+    /// if it were created with [`LinkedChunk::new_with_update_history`].
+    pub fn with_update_history(&mut self) {
+        self.build_with_update_history = true;
+    }
+
+    /// Run all error checks before reconstructing the full linked chunk.
+    ///
+    /// Must be called after checking `self.chunks` isn't empty.
+    ///
+    /// Returns the identifier of the first chunk.
+    fn check_consistency(&mut self) -> Result<ChunkIdentifier, LinkedChunkBuilderError> {
+        // Look for the first id.
+        let mut first_id = None;
+
+        for (id, chunk) in self.chunks.iter() {
+            if chunk.previous.is_none() {
+                // This chunk is a good candidate to be the first chunk.
+                first_id = Some(*id);
+                break;
+            }
+        }
```

Comment:
Nitpicking: This could have been a nice candidate for a:

```suggestion
        let first_id = self.chunks.iter().find_map(|(id, chunk)| chunk.previous.is_none().then_some(*id));
```